modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/disk/SortedExternalResult.java

```diff
@@ -0,0 +1,473 @@
+     */
+    private void addRowToHashIndex(Value[] row, long rowPosInFile) {
+        ValueRow distKey = getRowKey(row);
+
+        hashIdx.put(distKey, rowPosInFile);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void reset() {
+        spillRowsBufferToDisk();
+
+        if (resQueue != null) {
+            resQueue.clear();
+
+            for (Chunk chunk : chunks)
+                chunk.reset();
+        }
+        else {
+            resQueue = sort == null ? new LinkedList<>() : new PriorityQueue<>(new Comparator<Chunk>() {
+                @Override public int compare(Chunk o1, Chunk o2) {
+                    return sort.compare(o1.currentRow(), o2.currentRow());
+                }
+            });
+        }
+
+        // Init chunks.
+        for (Chunk chunk : chunks) {
+            if (chunk.next())
+                resQueue.offer(chunk);
+        }
+    }
+
+
+
+    /** {@inheritDoc} */
+    @Override protected void onClose() {
+        super.onClose();
+
+        U.closeQuiet(hashIdx);
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean contains(Value[] values) {
+        return getPreviousRow(values) != null;
+    }
+
+    /** {@inheritDoc} */
+    @Override public synchronized ResultExternal createShallowCopy() {
+        onChildCreated();
+
+        return new SortedExternalResult(this);
+    }
+
+    /**
+     * @return whether this result is a distinct result
+     */
+    private boolean isAnyDistinct() {
+        return distinct || distinctIndexes != null;
+    }
+
+    /**
+     * Extracts distinct row key from the row.
+     * @param row Row.
+     * @return Distinct key.
+     */
+    ValueRow getRowKey(Value[] row) {
+        if (distinctIndexes != null) {
+            int cnt = distinctIndexes.length;
+
+            Value[] newValues = new Value[cnt];
+
+            for (int i = 0; i < cnt; i++)
+                newValues[i] = row[distinctIndexes[i]];
+
+            row = newValues;
+        } else if (row.length > visibleColCnt)
+            row = Arrays.copyOf(row, visibleColCnt);
+
+        return ValueRow.get(row);
+    }
+
+    /**
+     * Sorted rows chunk on the disk.
+     */
+    private class Chunk {
+        /** Start chunk position. */
+        private final long start;
+
+        /** End chunk position. */
+        private final long end;
+
+        /** Current position within the chunk */
+        private long curPos;
+
+        /** Current row. */
+        private Value[] curRow;
+
+        /**
+         * @param start Start position.
+         * @param end End position.
+         */
+        Chunk(long start, long end) {
+            this.start = start;
+            this.curPos = start;
+            this.end = end;
+        }
+
+        /**
+         * @return {@code True} if next row is available within a chunk.
+         */
+        boolean next() {
+            while (curPos < end) {
+                curRow = readRowFromFile(curPos);
```

Comment:
Removed entry will be counted as the end of file, right?

Reply:
No, removed entries will be skipped until we reach non removed entry or the end of chunk.